File: lib/solvers/UselessViaRemovalSolver/break-route-into-sections.ts
import type { HighDensityRoute } from "lib/types/high-density-types";
import type { RouteSection } from "./route-section";

export const breakRouteIntoSections = (
  route: HighDensityRoute,
): RouteSection[] => {
  const routeSections: RouteSection[] = [];
  const routePoints = route.route;
  if (routePoints.length === 0) return [];

  let currentSection: RouteSection = {
    startIndex: 0,
    endIndex: -1,
    z: routePoints[0].z,
    points: [routePoints[0]],
  };
  for (let i = 1; i < routePoints.length; i++) {
    if (routePoints[i].z === currentSection.z) {
      currentSection.points.push(routePoints[i]);
    } else {
      currentSection.endIndex = i - 1;
      routeSections.push(currentSection);
      currentSection = {
        startIndex: i,
        endIndex: -1,
        z: routePoints[i].z,
        points: [routePoints[i]],
      };
    }
  }
  currentSection.endIndex = routePoints.length - 1;
  routeSections.push(currentSection);

  return routeSections;
};

File: rust/trace-simplification/src/useless_via_removal_solver/break_route_into_sections.rs
use crate::types::RouteRef;
use super::route_section::RouteSection;

pub fn break_route_into_sections(route: &RouteRef) -> Vec<RouteSection> {
    let route = route.borrow();
    let points = &route.route;
    if points.is_empty() { return Vec::new(); }
    let mut sections = Vec::new();
    let mut current = RouteSection { identity: crate::types::next_identity(), points_array_identity: crate::types::next_identity(),
        start_index: 0, end_index: -1, z: points[0].borrow().z,
        points: vec![points[0].clone()],
    };
    for (index, point) in points.iter().enumerate().skip(1) {
        if point.borrow().z == current.z {
            current.points.push(point.clone());
        } else {
            current.end_index = index as isize - 1;
            sections.push(current);
            current = RouteSection { identity: crate::types::next_identity(), points_array_identity: crate::types::next_identity(),
                start_index: index, end_index: -1, z: point.borrow().z,
                points: vec![point.clone()],
            };
        }
    }
    current.end_index = points.len() as isize - 1;
    sections.push(current);
    sections
}
